### ISSUE TITLE
How to replay retargeted .pkl data in Mujoco or Isaac Gym?

### ISSUE THREAD

**Issue by learning-github-zgy** [sjtu-mvasl-robotics/GBC#3]
Hello,

First of all, thank you for sharing this great work and for maintaining the repository.

I was able to run the pipeline and obtain the retargeted .pkl files. The trajectory comparison plots are helpful, but I would like to replay the motions directly inside a simulator (e.g., Mujoco or Isaac Gym) to better visualize the retargeted results.

I believe such a visualization would be very useful for verifying the quality of retargeting, beyond trajectory plots.

Thank you very much for your help!

**Comment by learning-github-zgy** [comment]
https://github.com/user-attachments/assets/3c8fb40c-4ddc-407a-b065-99887cc86ceb

Hi, I followed the PBHC repository and wrote a visualization script to replay the retargeted actions. While I can see the general motion, there are some issues:

1. **Foot penetration**: During walking, the feet sometimes penetrate the ground, indicating that the foot orientation is not properly preserved.
2. **Hand-leg collisions**: The hands occasionally intersect with the legs, which is unnatural and may indicate mapping or IK issues.
3. The overall motion is visually correct in trajectory but less accurate when directly replayed in Mujoco.

Would you be able to provide guidance on how to correctly replay the retargeted PKL files in Mujoco or Isaac Gym to avoid these issues?

Thanks a lot!

**Comment by securityCFS** [comment]
## Foot Penetration issues
For the foot penetration problems, now there are three kinds of solutions, as proposed in our paper:
* 1. **Additional Retargeting Joints**: The penetration problem is mostly caused by lacking of link position between ankle and toe (you can see from URDF file that that the origin of ankle links are the same). So as we did in the Turin's configuration, which shows promising results, that users can add a virtual joint at the top of the toe for better retargeting (this should be used in training setup)
* 2.  **Direct Mapping Loss**: This is a simpler approach, instead of adding joints, some joints can be trained to "mirror" the input euler angles from given MoCap angle-axis data. You can check the following codes:
https://github.com/sjtu-mvasl-robotics/GBC/blob/162c0e26c68f308666e0b3e9f98b430d004d3936/GBC/utils/data_preparation/pose_transformer_trainer.py#L399C5-L432C20
This loss function is enabled when a `direct mapping table` is parsed to `poseformer trainer`. A typical direct mapping looks like:
```python
joint_mapping_table = {
        "L_Ankle": {
            "pitch": "left_ankle_pitch_joint",
            "roll": "left_ankle_roll_joint",
        },
        "R_Ankle": {
            "pitch": "right_ankle_pitch_joint",
            "roll": "right_ankle_roll_joint",
        },
```

If you prefer not to retrain, you an also override the IK network output in the converter with
```python
converter.set_direct_joint_map_body(joint_mapping_table)
```
The following code shows how this override works in the converting process:
https://github.com/sjtu-mvasl-robotics/GBC/blob/162c0e26c68f308666e0b3e9f98b430d004d3936/GBC/utils/data_preparation/amass_action_converter.py#L421C13-L427C37

* 3. (Most Recommended) **Use Reinforcement Learning (RL)**:
Directly retargeting human mocap data is often imperfect, as the resulting motions may not be dynamically feasible for a humanoid robot. Therefore, we strongly recommend using our RL environment to learn how to execute these motions realistically.

By interacting with the physics-based environment, the agent learns to perform the retargeted actions in a dynamically feasible manner. We guide this learning process by providing foot-contact information and using reward functions that encourage stable foot placement (e.g., keeping feet parallel to the ground).

## Self-collision issues
That's an excellent point regarding self-collision between the hands and legs. Thank you for the great feedback!

The current open-source release only covers full-body retargeting. We do have supplementary code and guides for hand retargeting, but we are still working on fixing some bugs. We plan to release them in a future update. The 29dof version of Unitree G1 have fixed hand directions, our model doesn't change it.

Avoiding self-collision is a key improvement we're planning for the project's retargeting module and will be released with our diffusion project (~~~publishing articles always require making innovations and improvements, right? ~~~). Our proposed solution is to introduce an additional loss term, `self_collision_loss`, during the `PoseFormer` training phase.

However, a prerequisite for this loss is a simple MLP classifier trained to predict whether a given motion will result in self-collision (Using loss is to ensure we get something to measure the gradient flow. You can also use entach+detach method to force add a gradient to the following function). **This classifier is not yet reliable enough, largely due to a data imbalance issue.** Specifically, when training a classifier on motions that exclude hand joints, true self-collision samples are very rare in the dataset, which makes it difficult to achieve reliable accuracy. And for most of the MoCap data, humans are not likely to  We will push this update to the public repository once we have addressed this and validated its performance.

If you'd like to experiment with this yourself, you can get started with the following detect script using `pybullet`:
```python
import pybullet as p
def batch_check_self_collision(robot_urdf_path, all_joint_angles):
    """
    Batch check self-collision for multiple sets of joint angles.

    Args:
        robot_urdf_path (str): Path to the robot URDF file.
        all_joint_angles (list of lists): A list containing multiple sets of joint angles.

    Returns:
        list of bool: Boolean list indicating whether each pose has collision.
    """
    results = []
    
    client = p.connect(p.DIRECT)
    robot_id = p.loadURDF(robot_urdf_path, useFixedBase=True)

    num_joints = p.getNumJoints(robot_id)
    movable_joint_indices = [i for i in range(num_joints) if p.getJointInfo(robot_id, i)[2] != p.JOINT_FIXED]

    for joint_angles in all_joint_angles:
        if len(joint_angles) != len(movable_joint_indices):
            p.disconnect()
            raise ValueError("Number of joint angles does not match the number of movable joints!")

        for i, joint_index in enumerate(movable_joint_indices):
            p.resetJointState(robot_id, joint_index, joint_angles[i])

        p.performCollisionDetection()
        contact_points = p.getContactPoints(bodyA=robot_id, bodyB=robot_id)
        
        results.append(len(contact_points) > 0)

    p.disconnect()
    
    return results
```
Training an MLP would be easy, so I won't include it here. Once a classifier is trained, you can add it to the `PoseFormer` part. Directly implementing the function above to the training step is quite slow, but you can still try it out.

~~~A simpler solution is to enable self collision in RL training but I haven't tried it yet~~~

**Comment by securityCFS** [comment]
And for the visualization, use https://github.com/sjtu-mvasl-robotics/GBC/blob/162c0e26c68f308666e0b3e9f98b430d004d3936/GBC/viewer/isaacsim_viewer.py to play in Isaac Sim. I haven't written a generalized MuJoCo preview script yet, I'll see if copilot can help me with that.

**Comment by learning-github-zgy** [comment]
Thank you very much for your detailed reply and guidance! I’ll follow your suggestions.

For visualization, I’ll also make use of the Isaac Sim viewer you pointed out. The link you shared is very helpful, and I’ll start experimenting with it soon.

I really appreciate the time you spent clarifying these points. I’ll try things out accordingly, and I’m also looking forward to your potential work on the GBC-MUJICA. That would be very useful for many of us!

Thanks again for your support!

**Comment by securityCFS** [comment]
Thanks for supporting our project

**Comment by learning-github-zgy** [comment]
By the way, I have one additional question: do you have any demo videos showing the retargeting results of your trained network? I am curious whether the issues I mentioned earlier also appeared in your experiments.

Specifically, I have been following the tutorial to train both the shape fitting and the PoseTransformer, and I would like to know how the retargeting results looked on G1 in your experiments. Did you observe similar problems after training?

**Comment by securityCFS** [comment]
For pure retargeting, **yes, it can happen** (this is just how mocap data happens to be). Here is the retargeted data replay for H1-2 (we are mainly using this one, but G1 has quite the same  #problems), however, with an additional link, you can see that the feet problem is mostly fixed, self collision with hand, though, still exists.

https://github.com/user-attachments/assets/19cb68f5-b647-4a79-b90f-c1e9be0c8f08

However, after RL training, this problem will be solved (this is an adaptive strategy developed by agents).

We will also release a delta network to adjust retargeting issues when we collect enough RL performed trajectories, just like H2O and its successors. Our problem is that, since we don't actually own a G1 robot, we can now only collect data from MuJoCo simulations, that is not enough to compensate for sim2real gap. Sad :(

**Comment by learning-github-zgy** [comment]
Thank you again for your detailed reply and for sharing the retargeted video. I have a further question regarding the training configuration.

I followed the configuration mentioned here: [https://github.com/sjtu-mvasl-robotics/GBC/issues/2#issuecomment-3208882157](https://github.com/sjtu-mvasl-robotics/GBC/issues/2#issuecomment-3208882157), and after training, the effect looks similar to the video I attached previously. I would like to clarify:

1. **About virtual joints**: Are the additional virtual joints introduced only for making the retargeted motion look closer to the source data (i.e., better visual alignment during replay)? Since the robots do not have these extra joints, I assume they can not be used directly for imitation learning or control. Is this understanding correct?

2. **About comparison with optimization-based methods**: Some common retargeting approaches, such as MINK-based optimization (e.g., GMR, PBHC), seem to produce more stable results — for example, the feet orientation does not collapse as severely as in the Pose Transformer case. Could you help me understand why optimization-based retargeting methods tend to perform better in such aspects compared to the transformer-based approach?

Thank you very much for your patience in answering these detailed questions!

**Comment by securityCFS** [comment]
Sure. The virtual joints are used only for retargeting (it is used to calculate certain position of the toe, it will not be used in other places).

For optimization based methods, take MINK-based ones as example, they treat the whole retargeting process as an optimization problem and take advantage of physics engine like MuJoCo. Ours are quite simple, we formulate this problem as a constraint optimization problem that can be solved directly using python/pytorch. The generated model can convert **any motion input**, not just from the dataset (for example, for later training diffusion with HumanML3D, the whole HumanAct12 is converted using trained checkpoint, the whole conversion of the dataset takes only few minutes). One significant advantage of this is: we run much much more faster, as shown in the following chart:

<img width="1392" height="305" alt="Image" src="https://github.com/user-attachments/assets/5aaa9906-6935-4718-a0e3-03ff6c07b14c" />

And this allows, although not currently included in released codes, a real real-time imitation framework to be integrated to robots and can run on Nvidia Orin carried on humanoids **with very little calculation resource occupied**.

However, GBC being less accurate than MINK, are mainly a result of lacking constraints. So this is why we included `use_reference_actions` flag in `PoseFormer Trainer`. By replacing the IK in `RobotKinematics`, you can train the retargeting network with MINK pre-calculated retargeted motions. This will only slow down the first epoch (required for whole dataset pre-calculation), but the others will remain the same.

If you have enough GPU capacity, integrating MINK in this step would provide better universal models. However, my experiments are mostly conducted on an RTX 3070ti, it simply can't support both calculation at the same time :(

**Comment by securityCFS** [comment]
And to add more to this, if you want to include MINK in the training step, it is recommended to set `sample_steps` of the model to 20 or higher to reduce the amount of time used. With reference motion, the transformer no longer needs large dataset to converge. I tried this out using Unitree Retargeting Dataset from LAFAN1 (Type: H1-2) that is introduced in Feb and got promising results. If you need an official integration of such method, I'll consider adding it to our next major update.

**Comment by learning-github-zgy** [comment]
Thank you for the detailed explanation! Just to clarify, do you mean that the issue of the toe orientation collapsing below the ground does **not** affect the downstream imitation learning process? In my current retargeting results, the toes are consistently tilted — for example, the right toe points downward at about 45°. I am wondering whether this might be related to an incorrect default foot angle setting.

Also, regarding your suggestion about integrating MINK-based retargeted motions during training: I am very much looking forward to your next update. It would be great if we could achieve both **higher retargeting quality** and **fast conversion speed** at the same time.

Finally, I have a question about the reinforcement learning (RL) strategy you mentioned earlier for improving retargeting quality. If we perform RL training on retargeted data within a simulation environment, wouldn’t the overall retargeting pipeline become significantly more complex? While this can indeed yield strong results, it seems that training policies for every trajectory might not scale well for building a more **general-purpose behavior controller**. Could you share your perspective on this trade-off?

**Comment by securityCFS** [comment]
The purpose of our RL+IL framework is to adapt the output of motion retargeting into executable commands for a robot controller. Our primary goal is to achieve real-time imitation, where the robot can instantly perform any given input motion. We have successfully trained policies for specific tasks like walking, running, and even complex martial arts, but achieving generalization to any arbitrary motion remains a significant challenge.

Essentially, RL compensates for the inaccuracies of the initial retargeting step. It transforms motions that may be kinematically plausible but not dynamically feasible into physically realistic, dynamics-aware control commands.

Furthermore, our framework learns from multiple motions simultaneously rather than focusing on one at a time. By adding a diverse mix of motions (e.g., walking, running, and jumping) to the training environment, the agent learns to generalize and capture common underlying patterns. This is facilitated by the data augmentation techniques implemented in `ref_buffer.py` and our `ref_observation_manager.py`.

This approach contributes to our ultimate goal: developing a universal controller for a wide range of embodied tasks. Our upcoming diffusion-based planner, which is also part of this vision, serves the same purpose. While this universal controller is still under active development and I can't reveal too many details, we are targeting a release by the end of this year (2025), likely accompanied by a new multi-task dataset. (It might be slightly delayed since I am about to graduate...).

**Comment by learning-github-zgy** [comment]
I really admire your vision of developing a universal controller capable of handling diverse tasks — that is truly exciting. The idea of combining RL and IL to compensate for retargeting inaccuracies and achieve dynamics-aware control feels both elegant and practical.

Recently, I also noticed some works on general-purpose controllers, such as *BeyondMimic*, which seem conceptually related to your approach, though your framework’s focus on online imitation and multi-task generalization appears particularly promising. I am very much looking forward to your upcoming diffusion-based planner and the new multi-task dataset you mentioned.

Thank you for generously sharing these insights — they are very inspiring for my own research.

**Comment by securityCFS** [comment]
Thanks again for your attention! I also admire these generalized controller works and am trying to learn from them (since we both implement training with Isaac Lab). Their reward function design and early stop methodologies are really inspiring, and new experiments we run using G1 confirms that G1 is a much better platform for imitating than the H1-2 we originally use. Our new design of RL algorithm now converges on G1 in only 1500-3000 epochs, after learning from these pioneers. I will update the G1 training code after cleaning them up.

**Comment by learning-github-zgy** [comment]
Indeed, compared with H1-2, G1 does seem to have clear advantages in imitating highly dynamic behaviors, possibly due to its smaller size and agility. BeyondMimic also shows that when tracking a single motion, only a relatively small number of training epochs is required to achieve good results. I am very much looking forward to your subsequent update on the G1 training code. Thank you again for sharing your insights and progress!
